tortoise/tortoise.go

```diff
@@ -764,9 +762,29 @@ func (t *turtle) onHareOutput(lid types.LayerID, bid types.BlockID) {
 		return
 	}
 	t.logger.With().Debug("on hare output", lid, bid, log.Bool("empty", bid == types.EmptyBlockID))
-	t.decided[lid] = struct{}{}
-	if bid != types.EmptyBlockID {
-		t.hareOutput[bid] = support
+	previous, exists := t.hareOutput[lid]
+	if !exists || (exists && previous != bid) {
+		t.hareOutput[lid] = bid
+		if lid.Before(t.minprocessed) && t.mode.isVerifying() && withinHdist(t.Config, lid, t.last) {
+			t.logger.With().Info("local opinion changed within hdist",
+				lid,
+				log.Stringer("verified", t.verified),
+				log.Stringer("previous", previous),
+				log.Stringer("new", bid),
+			)
+			t.verifying.resetWeights()
+
+			// if local opinion within hdist was changed about the layer
+			// that was already verified we need to revert that
+			t.verified = minLayer(t.verified, lid)
+			for target := t.verified; !target.After(t.processed); target = target.Add(1) {
```

Comment:
does this have the same effect as doing rerun from t.verified layer?
conceptually it's easier to understand that if local opinion changes about hare output for layer N, then we do a rerun from layer N.

Reply:
i probably don't understand this comment
opinion changed about layer N, goodness of all ballots that depend on this opinion is reevaluated and good weight is recomputed